### DISCUSSION TITLE
How can I emulate Milvus L2 Euclidian distance in Python?  I get different from milvus results dist = np.linalg.norm(embeding_to_search2 - embeding_to_search1)

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18863] [Q&A and General discussion] [answered]
How do I get exact match like in milvus? 
I get quit different results in milvus than in usual 
dist = np.linalg.norm(embeding_to_search2 - embeding_to_search1)

**Comment by martinenkoEduard** [comment]
How can I emulate it in python given 2 vectors?

**Comment by martinenkoEduard** [comment]
I can create temporary collection but is a dirty trick :)

**Comment by martinenkoEduard** [comment]
I tried https://milvus.io/docs/v2.1.x/calculate_distance.md

But I get error


    params = {
        "metric": "L2", 
        "dim": 512
    }
    dist = utility.calc_distance(
        vectors_left=embeding_to_search1[0], 
        vectors_right=embeding_to_search2[0], 
        params=params
    )


192.168.56.1 - - [26/Aug/2022 13:50:48] "POST /verifyFaces HTTP/1.1" 500 -
 -6.06627986e-02  2.87021007e-02 -3.37629803e-02 -3.35696787e-02
  4.48604114e-02 -2.82180086e-02  7.29281306e-02 -5.29637747e-02
 -1.17187211e-02  1.30101861e-02  3.29176560e-02  2.12958250e-02
  2.86199339e-02 -3.15977596e-02 -3.14065330e-02  2.66272929e-02] is illegal

**Comment by martinenkoEduard** [comment]
Got it working

    params = {
        "metric": "L2", 
        "dim": 512
    }
    l1 = embeding_to_search1.tolist()
    l2 = embeding_to_search2.tolist()

    dist = utility.calc_distance(
        vectors_left={"float_vectors": l1}, 
        vectors_right={"float_vectors": l2}, 
        params=params
    )

**Comment by yhmo** [comment, marked as answer]
The standard Euclidian distance is:
![Selection_079](https://user-images.githubusercontent.com/2282099/187147611-6099c1ee-83d0-4457-93ee-ba8c1e950637.png)


As we know, this formula has a square root calculation which can lead to poor performance. For similarity comparison, the square root calculation doesn't affect the result. So, the search engine only returns the value without square root calculation as the distance. 

The np.linalg.norm() returns a standard Euclidian distance. So their result is different.

**Reply by martinenkoEduard** [reply]
Can you please share simple example on how to emulate Milvus's Euclidian distance calculations without the square root?

**Reply by yhmo** [reply]
The Euclidian distance formula is easy, you can use Numpy to calculate it by only one line: https://stackoverflow.com/questions/1401712/how-can-the-euclidean-distance-be-calculated-with-numpy

You can redo a square calculation to the result.
